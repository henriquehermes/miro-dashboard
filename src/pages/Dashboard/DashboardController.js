import React, { useState, useEffect } from 'react';
import api from '../../config/api';
import DashboardComponent from './DashboardComponent';

export default function Dashboard({ history }) {
  const [books, setBooks] = useState([]);
  const [filterBooks, setFilterBooks] = useState([]);
  const [searchText, setSearchText] = useState('');
  const [expand, setExpand] = useState(false);
  const [isVisible, setIsVisible] = useState(false);
  const [editBook, setEditBook] = useState(false);

  const isAdmin = localStorage.getItem('admin');

  useEffect(() => {
    async function loadBooks() {
      const response = await api.get('/books');
      setBooks(response.data);
      setFilterBooks(response.data);
    }
    loadBooks();
  }, []);

  function expandBook(index) {
    setExpand(index);
  }

  function filterList(value) {
    setSearchText(value);
    let items = books;
    items = items.filter(item => {
      return (
        item.title
          .toLowerCase()
          .trim()
          .search(searchText.toLowerCase()) !== -1
      );
    });
    setFilterBooks(items);
  }

  function handleLogout() {
    localStorage.clear();
    history.push('/');
  }

  async function rentBook(book) {
    book.rented = true;
    await api.put(`/books/${book.id}`, book);

    const { data } = await api.get('/books');
    setBooks(data);
    setFilterBooks(data);
  }

  function handleUpdate(newBook, isChanged) {
    if (isChanged) {
      const newList = books.map(b => {
        if (b.id === newBook.id) b = newBook;
        return b;
      });

      setBooks(newList);
      setFilterBooks(newList);
    } else {
      const newList = books;
      newList.push(newBook);

      setBooks(newList);
      setFilterBooks(newList);
    }
  }

  async function handleDelete(oldBook) {
    const newList = books.filter(b => {
      if (b.id !== oldBook.id) return b;
    });

    await api.delete(`/books/${oldBook.id}`);

    setBooks(newList);
    setFilterBooks(newList);
  }

  return (
    <DashboardComponent
      handleUpdate={(newBook, isChanged) => handleUpdate(newBook, isChanged)}
      handleDelete={oldBook => handleDelete(oldBook)}
      rentBook={book => rentBook(book)}
      handleLogout={handleLogout}
      filterList={value => filterList(value)}
      expandBook={index => expandBook(index)}
      isAdmin={isAdmin}
      filterBooks={filterBooks}
      expand={expand}
      searchText={searchText}
      isVisible={isVisible}
      editBook={editBook}
      setEditBook={setEditBook}
      setIsVisible={setIsVisible}
    />
  );
}
