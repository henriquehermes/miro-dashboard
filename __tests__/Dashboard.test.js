import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import DashboardController from '../src/pages/Dashboard/DashboardController';

const theme = {
  primary: '#00bb77',
  secondary: '#282a36',
};

describe('Dashboard', () => {
  test('Render Dashboard', () => {
    const wrapper = renderer
      .create(<DashboardController theme={theme} />)
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });
});
