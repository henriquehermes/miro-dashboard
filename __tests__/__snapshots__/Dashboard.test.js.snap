// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Render Dashboard 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  padding: 40px;
}

.c2 {
  background: #f6f6f6;
  padding: 20px 16px;
  border-radius: 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 40px;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c6 {
  background: #f6f6f6;
  margin-top: 20px;
  padding: 20px 16px;
  border-radius: 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 {
  color: #282a36;
  font-weight: bold;
  margin-bottom: 20px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 45px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  margin-bottom: 10px;
}

.c3 {
  height: 40px;
  width: 100%;
  margin: 10px 0;
  padding: 0 10px;
  border: none;
  background: #fff;
  font-weight: bold;
}

.c4 {
  cursor: pointer;
  height: 40px;
  width: 90px;
  font-weight: bold;
  background: #fff;
  border-radius: 6px;
  border-color: #fff;
  border-style: solid;
  box-shadow: 0px 4px 18px 0px rgba(0,0,0,0.1);
}

.c5 {
  cursor: pointer;
  height: 40px;
  width: 90px;
  font-weight: bold;
  background: #fff;
  border-radius: 6px;
  border-color: #fff;
  border-style: solid;
  box-shadow: 0px 4px 18px 0px rgba(0,0,0,0.1);
  margin: 0px 5px;
}

@media (max-width:767px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    height: auto;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <input
        className="c3"
        id="search"
        onChange={[Function]}
        placeholder="Informe o titulo do livro"
        type="text"
        value=""
      />
      <button
        className="c4"
        onClick={[Function]}
      >
        Pesquisar
      </button>
    </div>
    <button
      className="c5"
      onClick={[Function]}
    >
      Sair
    </button>
  </div>
  <div
    className="c6"
  >
    <h1
      className="c7"
    >
      Lista de livros
    </h1>
    <div>
      <h1>
        Nenhum livro cadastrado :(
      </h1>
    </div>
  </div>
</div>
`;
